idmtest/idmtest.go

```diff
@@ -284,6 +258,85 @@ func (srv *Server) user(name string) *user {
 	return srv.users[name]
 }
 
+func (srv *Server) checkThirdPartyCaveat(ctx context.Context, req *http.Request, info *bakery.ThirdPartyCaveatInfo, token *httpbakery.DischargeToken) ([]checkers.Caveat, error) {
+	if srv.defaultUser != "" {
+		return []checkers.Caveat{
+			idmclient.UserDeclaration(srv.defaultUser),
+		}, nil
+	}
+	dischargeID := base64.URLEncoding.EncodeToString(info.Caveat)
+	if token == nil || token.Kind != "agent" {
+		ierr := httpbakery.NewInteractionRequiredError(nil, req)
+		agent.SetInteraction(ierr, "/login/agent?discharge-id="+dischargeID)
+		return nil, ierr
+	}
+	var ms macaroon.Slice
+	if err := ms.UnmarshalBinary(token.Value); err != nil {
+		return nil, errgo.Mask(err)
+	}
+	ops, _, err := srv.Bakery.Oven.MacaroonOps(ctx, ms)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	username := ""
+	for _, op := range ops {
+		if strings.HasPrefix(op.Entity, "user-") && op.Action == "discharge" {
+			username = strings.TrimPrefix(op.Entity, "user-")
+			break
+		}
+	}
+	_, err = srv.Bakery.Checker.Auth(ms).Allow(
+		ctx,
+		bakery.Op{
+			Entity: "discharge-" + dischargeID,
+			Action: "discharge",
+		},
+		bakery.Op{
+			Entity: "user-" + username,
+			Action: "discharge",
+		},
+	)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	return []checkers.Caveat{
+		idmclient.UserDeclaration(username),
+	}, nil
+}
+
+func (srv *Server) newHandler(p httprequest.Params, req interface{}) (*handler, context.Context, error) {
+	if err := srv.checkLogin(p.Context, p.Request, req); err != nil {
+		return nil, nil, errgo.Mask(err, errgo.Any)
+	}
+	return &handler{srv}, p.Context, nil
+}
+
+func (srv *Server) checkLogin(ctx context.Context, req *http.Request, treq interface{}) error {
+	var ops []bakery.Op
+	switch treq.(type) {
+	case *agentMacaroonRequest:
+		return nil
+	case *groupsRequest:
+		ops = append(ops, bakery.LoginOp)
```

Comment:
The "standard" way to do this would not be to explicitly include LoginOp here, but to implement a function that maps request to the operations required for that request (none of which would include LoginOp), and then use an ACLAuthorizer that returns the ACL for any given operation. When a request arrives without any macaroons, the bakery.Checker would use the IdentityClient implementation to return a discharge-required error ("i need an identity") and the code would see that and return a LoginOp macaroon in a discharge-required error to the client.

See blues-identity/internal/v1/auth.go and blues-identity/internal/auth/auth.go.

It's probably worth doing.